### PR TITLE
nvhpc: new package version 20.7

### PR DESCRIPTION
**Pull request by ajkotobi**
Nvidia HPC SDK package

Closes #19139

### CODE REVIEW

File: var/spack/repos/builtin/packages/nvhpc/package.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+       libraries and software tools essential to maximizing developer productivity
+       and the performance and portability of HPC applications.
+    """
+
+    homepage = "https://developer.nvidia.com/hpc-sdk"
+    url      = "https://developer.download.nvidia.com/hpc-sdk/nvhpc_2020_207_Linux_x86_64_cuda_multi.tar.gz"
+
+    maintainers = ['ajkotobi']
+
+    version('2020_207_Linux_x86_64_cuda_multi', sha256='a5c5c8726d2210f2310a852c6d6e03c9ef8c75e3643e9c94e24909f5e9c2ea7a')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/nvhpc/package.py]: I would probably call this version `2020_207`. You can add a `url_for_version` to help Spack figure out the URL

File: var/spack/repos/builtin/packages/nvhpc/package.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    variant(
+        'cuda', default='11.0', description='List of CUDA that are enabled',
+        values=('11.0', '10.2', '10.1'), multi=False
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/nvhpc/package.py]: You might want to consider extending the `CudaPackage` base class. It adds a `cuda_arch` variant that can be used across packages.

File: var/spack/repos/builtin/packages/nvhpc/package.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+import os
+
+
+class Nvhpc(Package):
+    """The NVIDIA HPC Software Development Kit (SDK) includes the proven compilers,
+       libraries and software tools essential to maximizing developer productivity
+       and the performance and portability of HPC applications.
```
Comment (pramodk) [inline var/spack/repos/builtin/packages/nvhpc/package.py]: `remark` : as this is NVIDIA's hpc-sdk, I was wondering if this should be available to users as `nv-hpc-sdk` or `nvidia-hpc-sdk`.

**Comment by fspiga** [comment]
Hello! NVIDIA HPC SDK supports also ppc64le (POWER) and aarch64 (Arm). It will be really useful to include the support for these two architecture in the Spack package as well.

**Comment by ajkotobi** [comment]
@fspiga , I will add it to this PR.

**Comment by ajkotobi** [comment]
Going to fix/complete this PR to avoid future duplication.

File: var/spack/repos/builtin/packages/nvhpc/package.py (hunk 2 above)
Comment (ajkotobi) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: This added to PR

File: var/spack/repos/builtin/packages/nvhpc/package.py (hunk 1 above)
Comment (ajkotobi) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: This one fixed and moved to `url_for_version`

File: var/spack/repos/builtin/packages/nvhpc/package.py (hunk 3 above)
Comment (ajkotobi) [reply var/spack/repos/builtin/packages/nvhpc/package.py]: its name changed to `nvidia-hpc-sdk`

**Comment by ajkotobi** [comment]
@adamjstewart , could you review this again?

File: var/spack/repos/builtin/packages/nvidia-hpc-sdk/package.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+          values=('11.0', '10.2', '10.1'), multi=False)
+
+    variant(
+        'network', default='network', description='Network installation',
+        values=('network', 'single'), multi=False
+    )
+
+    depends_on('cuda@11.0.2:',   when='cuda=11.0')
+    depends_on('cuda@10.2.89:',  when='cuda=10.2')
+    depends_on('cuda@10.1.243:', when='cuda=10.1')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/nvidia-hpc-sdk/package.py]: With regards to the CUDA stuff, I would remove the variant and instead use:
```python
depends_on('cuda')
conflicts('^cuda@:10', when='platform=aarch64')
```
This allows the user to control the version of CUDA used from the command-line, like:
```console
$ spack install nvidia-hpc-sdk ^cuda@10.2.89
```

File: var/spack/repos/builtin/packages/nvidia-hpc-sdk/package.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+           url = "https://developer.download.nvidia.com/hpc-sdk/nvhpc_{0}_Linux_{1}_cuda_11.0.tar.gz"
+        else:
+           url = "https://developer.download.nvidia.com/hpc-sdk/nvhpc_{0}_Linux_{1}_cuda_multi.tar.gz"
+        return url.format(version, platform.machine())
+
+    def install(self, spec, prefix):
+
+        os.environ['NVHPC_SILENT'] = "true"
+        os.environ['NVHPC_INSTALL_DIR'] = self.prefix
+        os.environ['NVHPC_DEFAULT_CUDA'] = self.spec.variants['cuda'].value
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/nvidia-hpc-sdk/package.py]:
```suggestion
        os.environ['NVHPC_DEFAULT_CUDA'] = str(self.spec['cuda'].version.up_to(2))
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Review by adamjstewart — APPROVED** [review]

**Comment by adamjstewart** [comment]
Flake8:
```
var/spack/repos/builtin/packages/nvidia-hpc-sdk/package.py:14: [E501] line too long (82 > 79 characters)
var/spack/repos/builtin/packages/nvidia-hpc-sdk/package.py:24: [E128] continuation line under-indented for visual indent
var/spack/repos/builtin/packages/nvidia-hpc-sdk/package.py:31: [E111] indentation is not a multiple of four
var/spack/repos/builtin/packages/nvidia-hpc-sdk/package.py:33: [E111] indentation is not a multiple of four
var/spack/repos/builtin/packages/nvidia-hpc-sdk/package.py:40: [E501] line too long (82 > 79 characters)
var/spack/repos/builtin/packages/nvidia-hpc-sdk/package.py:49: [E226] missing whitespace around arithmetic operator
var/spack/repos/builtin/packages/nvidia-hpc-sdk/package.py:49: [E226] missing whitespace around arithmetic operator
var/spack/repos/builtin/packages/nvidia-hpc-sdk/package.py:51: [E128] continuation line under-indented for visual indent
var/spack/repos/builtin/packages/nvidia-hpc-sdk/package.py:51: [E226] missing whitespace around arithmetic operator
var/spack/repos/builtin/packages/nvidia-hpc-sdk/package.py:51: [E226] missing whitespace around arithmetic operator
var/spack/repos/builtin/packages/nvidia-hpc-sdk/package.py:53: [E226] missing whitespace around arithmetic operator
var/spack/repos/builtin/packages/nvidia-hpc-sdk/package.py:53: [E226] missing whitespace around arithmetic operator
```

**Comment by pramodk** [comment]
@ajkotobi  : could you take take final look to fix flake8 warnings so that this can be merged?

Edit : Also, #19139 takes of power and arm architectures with mpi variant. So I would suggest to merge good parts from #19139 in this PR.

**Comment by ajkotobi** [comment]
> @ajkotobi : could you take take final look to fix flake8 warnings so that this can be merged?
> 
> Edit : Also, #19139 takes of power and arm architectures with mpi variant. So I would suggest to merge good parts from #19139 in this PR.

Today going to wrap the flake8 issue up to finalize this PR.

**Comment by adamjstewart** [comment]
If you ever have trouble with the flake8 stuff, you can run `spack flake8` if flake8 is installed and it will tell you what needs to be changed.

**Pull request closed without merging**